pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java

```diff
@@ -0,0 +1,168 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.rebalance;
+
+import com.google.common.base.Preconditions;
+import java.util.Map;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.spi.utils.CommonConstants;
+/**
+ * <code>ZkBasedTableRebalanceObserver</code> observes rebalance progress and tracks rebalance status,
+ * stats in Zookeeper. This will be used to show the progress of rebalance to users via rebalanceStatus API.
+ */
+public class ZkBasedTableRebalanceObserver implements TableRebalanceObserver {
+    private final String _tableNameWithType;
+    private final String _rebalanceJobId;
+    private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+    TableRebalanceProgressStats _tableRebalanceProgressStats;
+
+    public ZkBasedTableRebalanceObserver(String tableNameWithType,
+                                         String rebalanceJobId,
+                                         ZkHelixPropertyStore<ZNRecord> propertyStore) {
+        Preconditions.checkState(tableNameWithType != null, "Table name cannot be null");
+        Preconditions.checkState(rebalanceJobId != null, "rebalanceId cannot be null");
+        Preconditions.checkState(propertyStore != null, "propertyStore cannot be null");
+        _tableNameWithType = tableNameWithType;
+        _rebalanceJobId = rebalanceJobId;
+        _propertyStore = propertyStore;
+        _tableRebalanceProgressStats = new TableRebalanceProgressStats();
+    }
+
+    @Override
+    public void onTrigger(Trigger trigger, Map<String, Map<String, String>> currentState,
+                          Map<String, Map<String, String>> targetState) {
+        switch (trigger) {
+            case START_TRIGGER:
+                updateOnStart(currentState, targetState);
+                break;
+            case IDEAL_STATE_CHANGE_TRIGGER:
+                _tableRebalanceProgressStats.setCurrentToTargetConvergence(getDifferenceBetweenTableRebalanceStates(
+                        targetState, currentState));
+                break;
+            case EXTERNAL_VIEW_TO_IDEAL_STATE_CONVERGENCE_TRIGGER:
+                _tableRebalanceProgressStats.setExternalViewToIdealStateConvergence(
+                        getDifferenceBetweenTableRebalanceStates(targetState, currentState));
+                break;
+            default:
+                throw new IllegalArgumentException("Unimplemented trigger: " + trigger);
+        }
+        PinotHelixResourceManager.addRebalanceResultToZk(_tableNameWithType, _rebalanceJobId,
```

Comment:
cc: @Jackie-Jiang ?

Reply:
Yes, Jackie and I discussed about this. We were thinking of using diff or a timer for these updates.  Using the in-memory state, gets rid of all these intermediate updates and will make this deterministic (one update per rebalance). I've incorporated that in my latest PR update. The only concern I had was that we need to ensure that this state is cleared and does not hang around. I've  added assertions to make sure that the in-memory is cleared.

Reply:
@saurabhd336 and I discussed. The in-memory state is local to the controller and will not be available to other controllers. We'll have to query all controllers to figure out which one has the rebalanceStatus. This approach complicates the status API as well as keep around in-memory state (HashMap). We decided to take the other approach to persist in ZK, but only when there's a difference in the stats computed. This will minimize updates to Zk. Added stats to track number of updates to Zk so that we'll know the write load on Zk.